ambry-server/src/main/java/com.github.ambry.server/AmbryServer.java

```diff
@@ -216,6 +216,7 @@ void shutdown() {
           logger.error("Error while closing notification system.", e);
         }
       }
+      clusterMap.close();
```

Comment:
Didn't realize that `shutdown()` could be called even if `startup()` did not succeed (which is a bit strange to begin with - ideally startup() should cleanup before returning a failure). I have added documentation for the shutdown methods in both the `RestServer` and `AmbryServer` to reflect this fact. Have also added the `null` checks.

And the factory creates agents only when asked for. This is particularly important with the participants (which will not be created by the frontends). The factories are also singletons - that is they only construct each type of agent, once. I have added this to the `ClusterAgentsFactory` interface.